src/main/java/org/truffleruby/language/globals/GlobalVariableStorage.java

```diff
@@ -26,18 +28,25 @@
 
     private volatile Object value;
 
+    private final Object defaultValue;
     private final DynamicObject getter;
     private final DynamicObject setter;
 
     GlobalVariableStorage(Object value, DynamicObject getter, DynamicObject setter) {
-        this.value = value;
+        this.defaultValue = value;
+        this.value = UNSET_VALUE;
         this.getter = getter;
         this.setter = setter;
         assert (getter != null) == (setter != null);
     }
 
     public Object getValue() {
-        return value;
+        Object currentValue = value;
+        return currentValue == UNSET_VALUE ? defaultValue : currentValue;
```

Comment:
This logic feels complex to me.
Can we not just create GlobalVariableStorage with `value == UNSET_VALUE` in GlobalVariables?
Also, can you add specs to illustrate what you are trying to fix?

Reply:
I'll add specs. Essentially the problem is that `defined? $my_var` should be false if `$my_var` has not been assigned, and `$my_var` should return the default value which is normally `nil`. However if `$my_var` has anything assigned to it (even the default value) then it becomes defined. We need to correctly distinguish between those states.

Reply:
I see.
Another way to fix this would be to not create the storage on reading `$my_var` but that's more fragile and complicated.
You could also keep a simple boolean to note if the variable was ever written to.

Reply:
I like the boolean idea in theory, but it requires two volatile reads on every access, or more specialisation on ReadGlobalVariableNode and some extremely careful checking in a few other places.